src/snowcli/cli/nativeapp/artifacts.py

```diff
@@ -0,0 +1,175 @@
+import os
+import shutil
+from pathlib import Path
+from typing import List, Union
+from dataclasses import dataclass
+
+
+class ArtifactError(Exception):
+    """
+    Common base class for all artifact errors.
+    """
+
+    pass
+
+
+class GlobMatchedNothingError(ArtifactError):
+    """
+    No files were found that matched the provided glob pattern.
+    """
+
+    def __str__(self):
+        return self.__doc__
+
+
+class SourceNotFoundError(ArtifactError):
+    """
+    The specifically-referenced source file or directory was not found
+    in the project directory.
+    """
+
+    path: Path
+
+    def __init__(self, path: Path):
+        super().__init__()
+        self.path = path
+
+    def __str__(self):
+        return self.__doc__ + f"\npath = {self.path}"
+
+
+class TooManyFilesError(ArtifactError):
+    """
+    Multiple files were mapped to one output file.
+    """
+
+    dest_path: Path
+
+    def __init__(self, dest_path: Path):
+        super().__init__()
+        self.dest_path = dest_path
+
+    def __str__(self):
+        return self.__doc__ + f"\ndest_path = {self.dest_path}"
+
+
+class OutsideDeployRootError(ArtifactError):
+    """
+    The specified path is outside of the deploy root.
+    This can happen when a relative path with ".." is provided.
+    """
+
+    dest_path: Path
+    deploy_root: Path
+
+    def __init__(self, dest_path: Path, deploy_root: Path):
+        super().__init__()
+        self.dest_path = dest_path
+        self.deploy_root = deploy_root
+
+    def __str__(self):
+        return (
+            self.__doc__
+            + f"\ndest_path = {self.dest_path}"
+            + f"\ndeploy_root = {self.deploy_root}"
+        )
+
+
+@dataclass
+class SrcDestPair:
```

Comment:
Can we name it more meaningfully?  Without context I don't understand "A pair of paths, relative to some context."